### PR TITLE
add accept num, emit num metric for ChainSpeculativeSampling

### PR DESCRIPTION
**Pull request by LiuXiaoxuanPKU**

### CODE REVIEW

File: include/flashinfer/sampling.cuh [hunk 1]
```diff
@@ -1175,10 +1178,28 @@ __global__ void ChainSpeculativeSampling(DType* draft_probs, IdType* draft_token
...
+          p = draft_probs[(row_idx * num_speculative_tokens + i) * d + draft_id];
+    DType u = uniform_samples[row_idx * (num_speculative_tokens + 1) + i];
+    if (u * p < q) {
+      ++accepted_token_num;
+    }
+  }
+
+  if (tx == 0) {
+    output_accepted_token_num[row_idx] = accepted_token_num;
+    output_emitted_token_num[row_idx] = emitted_token_num;
```
Comment (yzh119) [inline include/flashinfer/sampling.cuh]: Can we make it in-place update? If user specify an `output_accepted_token_num` array, we update its value directly, otherwise we return an a new array.

```suggestion
    output_accepted_token_num[row_idx] += accepted_token_num;
    output_emitted_token_num[row_idx] += emitted_token_num;
```

File: python/csrc/sampling.cu [hunk 2]
```diff
@@ -315,9 +315,11 @@ torch::Tensor top_k_mask_logits(torch::Tensor logits, std::optional<torch::Tenso
...
 }
 
-torch::Tensor chain_speculative_sampling(torch::Tensor draft_probs, torch::Tensor draft_token_ids,
-                                         torch::Tensor uniform_samples, torch::Tensor target_probs,
-                                         bool deterministic) {
+std::vector<torch::Tensor> chain_speculative_sampling(torch::Tensor draft_probs,
+                                                      torch::Tensor draft_token_ids,
+                                                      torch::Tensor uniform_samples,
+                                                      torch::Tensor target_probs,
+                                                      bool deterministic) {
```
Comment (yzh119) [inline python/csrc/sampling.cu]:
```suggestion
                                                      std::optional<torch::Tensor> maybe_output_accepted_token_num,
                                                      std::optional<torch::Tensor> maybe_output_emitted_token_num,
                                                      bool deterministic) {
```

File: python/csrc/sampling.cu [hunk 3]
```diff
@@ -349,14 +351,20 @@ torch::Tensor chain_speculative_sampling(torch::Tensor draft_probs, torch::Tenso
   auto output_token_ids = torch::empty({batch_size, num_speculate_tokens + 1},
                                        torch::dtype(torch::kInt32).device(device));
 
+  auto output_accepted_token_num =
+      torch::zeros({batch_size}, torch::dtype(torch::kInt32).device(device));
+  auto output_emitted_token_num =
+      torch::zeros({batch_size}, torch::dtype(torch::kInt32).device(device));
```
Comment (yzh119) [inline python/csrc/sampling.cu]:
```suggestion
  auto output_accepted_token_num =
      maybe_output_accepted_token_num.value_or(torch::zeros({batch_size}, torch::dtype(torch::kInt32).device(device)));
  auto output_emitted_token_num =
      maybe_output_emitted_token_num.value_or(torch::zeros({batch_size}, torch::dtype(torch::kInt32).device(device)));
```

File: include/flashinfer/sampling.cuh [hunk 4]
```diff
@@ -1166,6 +1168,7 @@ __global__ void ChainSpeculativeSampling(DType* draft_probs, IdType* draft_token
       SamplingTempStorage<DType, BLOCK_THREADS, SCAN_ALGORITHM, REDUCE_ALGORITHM>&>(smem_sampling);
 
   uint32_t pos = 0;
+  uint32_t stop_pos = num_speculative_tokens;
   for (pos = 0; pos < num_speculative_tokens; ++pos) {
```
Comment (yzh119) [inline include/flashinfer/sampling.cuh]: I prefer rename the iteration variable to `i`, which is consistent with the later loop.

**Review by yzh119 — COMMENTED** [review]
Thanks for your contribution @LiuXiaoxuanPKU ! I added a few comments and suggestions.

**Comment by yzh119** [comment]
This file has to be updated: https://github.com/flashinfer-ai/flashinfer/blob/338b2f5b99ea4ec61fd90b1c3f4c0812bfb00e36/python/csrc/flashinfer_ops.h#L70-L72

if you change the function signature.

**Comment by LiuXiaoxuanPKU** [comment]
@yzh119 Thanks for the review. I just fixed the comments, feel free to take another pass. Thanks!

**Review by yzh119 — APPROVED** [review]
LGTM, thank you!
